### PR TITLE
🐛 Source Bing Ads : fix rollback window for marketing performance reports

### LINKED ISSUE #22930: Source Bing Ads - No rollback window for conversions
## Environment
- **Airbyte version**:  0.39.42-alpha
- **OS Version / Instance**: AWS EC2 (Amazon Linux)
- **Deployment**: Docker
- **Source Connector and version**: Bing Ads 0.1.16
- **Step where error happened**: connector behaviour

## Current Behavior
There are data discrepancies on a few fields between Bing Ads API and Bing Ads data ingested on our data warehouse (BigQuery) through the Airbyte stream `ad_group_performance_report_daily`. The concerned fields are related to _conversions_ marketing concept (see _References_ section to learn more about marketing conversions):

- `AllConversions`
- `Conversions`
- `ConversionsQualified`

The ingested data reported **fewer conversions** than the one reported on Bing Ads API.

## Expected Behavior
Conversion field values are accurate. They are the same as the one of the API.

## Logs
I don't have any relevant logs to provide. The job succeeds but its behavior is not the one expected.

## Steps to Reproduce
1. Setup Bing Ads source (_Replication start date_ set to `2022-01-01`)
2. Setup BigQuery destination (_Loading method_ set to `Standard Inserts` & _Transformation Query Run Type_ set to `batch`)
3. Setup the connector (_Replication frequency_ set to `24 hours`) by enabling stream `ad_group_performance_report_daily` in `Incremental | Deduped + history` mode with `TimePeriod` as cursor field (for the record, this field has a daily granularity)
4. Setup conversion window of 30 days on Bing Ads UI

## Explanations
I deep-dived into Bing Ads and Airbyte documentation to find the issue's root cause and I came up with an explanation for the unexpected behavior.

In short, the **cursor field** is not appropriately implemented as it doesn't have any **rollback window** to take into account conversions that could happen during the attribution window.

### Example

Initial state:
- Attribution window: 30 days
- `TimePeriod` cursor field value: `15/10/2022`

Event:
- A conversion related to `5/10/2022` happens on `20/10/2022`. As the conversion happens within the 30 days attribution window, it should be counted for the day `5/10/2022`. The `conversions` field value for the date `5/10/2022` should be incremented by 1 after the next job run.

Job run:
- Run date: `25/10/2022`
- As `TimePeriod` cursor field value is set `15/10/2022`, only data with a `TimePeriod` after `15/10/2022` (included) will be ingested. The rest will be skipped, e.g. data from `5/10/2022`. So the conversion related to `5/10/2022` which happened on the `20/10/2022` won't be counted.

### Conclusion

If a marketing conversion related to a `TimePeriod` before the cursor field value happens after the cursor field value, this conversion is currently not taken into account.

## Note

A similar conversion issue pattern has already been reported and fixed on Twilio Source. See the history of its [source.py file](https://github.com/airbytehq/airbyte/blob/d32cec811331e280a46dcb01fdeef153e8d86379/airbyte-integrations/connectors/source-twilio/source_twilio/source.py#L57) for more information.

## References
- [About Bing Ads conversions](https://help.ads.microsoft.com/#apex/ads/en/56710/2)
- [Airbyte incremental sync](https://docs.airbyte.com/connector-development/tutorials/cdk-tutorial-python-http/read-data/)
- [Twilio source connector with implemented rollback window](https://github.com/airbytehq/airbyte/blob/d32cec811331e280a46dcb01fdeef153e8d86379/airbyte-integrations/connectors/source-twilio/source_twilio/source.py#L57)

## Are you willing to submit a PR?
Yes.

### PR DESCRIPTION
**Pull request by tlatrace**
## What
Bing Ads connector doesn’t ingest all **conversions** for marketing performance reports. There are discrepancies between Bing UI data and data received in the Destination.

At the moment, the conversions' fields (_Conversions_, _AllConversions_, _ConversionsQualified_) in the Destination only count conversions that happened the day of the click. Conversions happening after 2+ days, during the attribution window, are not counted.

The problem comes from the connector **cursor field** (on the incremental sync) not being appropriately implemented. Indeed, it misses a **rollback window** to attribute conversions happening after 2+ days.

This PR intends to add an optional parameter `lookback_window`in the connector configuration, for performance reports ingested with incremental streams.

Fixes #22930.

## How
During a sync, instead of retrieving data from `[cursor_value, now()]`, we’ll retrieve data from `[cursor_value - rollback_window, now()]`. So the API call logic is modified (not the cursor logic).

This means that if the cursor value is set to 15/12/2022, we’ll retrieve data up to 15/11/2022 (for a 30-day attribution window).

Cons: we’ll ingest lots of duplicates in the history table (30 duplicates per record in most cases!). The `deduped + history` feature will handle the deduplication in the final table.

## Recommended reading order
Review per commit, in chronological order.

## 🚨 User Impact 🚨
No breaking changes as the `lookback_window` is an optional parameter.

## Pre-merge Checklist

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

</details>

### CODE REVIEW

**Comment by misteryeo** [comment]
The team's going to take a first look at this shortly @tlatrace! Assigning to @sajarin

**Review by sajarin — COMMENTED** [review]
Hey @tlatrace, thanks for the contribution. 

On a first glance, this looks good. A small nit would be to rename `_get_start_date` to simply `get_start_date` (without a prefixing underscore) so that it matches the naming conventions of the other methods. 

Additionally, could you explain what the test case you've added is testing?

File: airbyte-integrations/connectors/source-bing-ads/unit_tests/test_reports.py [hunk 1]
```diff
@@ -38,6 +38,14 @@ def test_get_column_value():
     assert test_report.get_column_value(record, "Spend") == 1.203
 
 
+def test_get_updated_state_init_state():
```
Comment (tlatrace) [inline airbyte-integrations/connectors/source-bing-ads/unit_tests/test_reports.py:41]: This unit test makes sure that the state is correctly initiated for a given stream.

Nothing related to the fix, but I took the opportunity to add it.

**Comment by tlatrace** [comment]
Here are screenshots of **unit tests** and **integration tests** passing.

unit-tests
integration-tests

**Comment by sajarin** [comment]
@tlatrace running the tests again, I'm not able to get integration tests passing locally given our credentials despite setting the lookback_window to a default value of 0. 

Do you screenshots of passing **integration** tests? The above images includes passing unit but for integration tests, I'm not seeing any tests being run there. Just trying to verify that this is working so that I can debug my local and get it working in the CI/CD.

**Comment by tlatrace** [comment]
Following the [documentation](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-bing-ads#integration-tests),  I first tried to run integration tests with this Docker command:

```
docker build . --no-cache -t airbyte/source-bing-ads:dev && python -m pytest -p connector_acceptance_test.plugin
```

But I reached this error for 4 tests: `FileNotFoundError: [Errno 2] No such file or directory: '/data/tap_config.json'`, which I don’t understand and was already reported [here](https://airbytehq.slack.com/archives/C027KKE4BCZ/p1668684350063869). These tests also fails with the same error on the `master` branch.

I then tried to run integration tests using Gradle:

```
./gradlew :airbyte-integrations:connectors:source-bing-ads:integrationTest
```

Which led to `Timeout` errors on the 2 tests failing on the CI. These tests also fail on `master` branch on my local FYI. Reducing the number of streams in the `configured_catalog.json` doesn't help.

Could you please provide help on this? 🙏 

Traceback: [logs-bing-ads-source-integration-tests.txt](https://github.com/airbytehq/airbyte/files/11021115/logs-bing-ads-source-integration-tests.txt)

**Review by marcosmarxm — APPROVED** [review]
Thanks @tlatrace for the contribution. I'm going to finish the PR https://github.com/airbytehq/airbyte/pull/25259 because I don't have commit access to your company fork.

**Comment by tlatrace** [comment]
Thank you for your feedback @marcosmarxm. Just to be sure, is something else required from my side?

**Comment by marcosmarxm** [comment]
> Thank you for your feedback @marcosmarxm. Just to be sure, is something else required from my side?

Nope! Right now I'm going to wait the final approval from the connector team. Because Bing Ads it is a GA connector there are some discussion to merge new code into the connector.

**Comment by marcosmarxm** [comment]
@tlatrace tests are been running in https://github.com/airbytehq/airbyte/pull/25259

**Comment by marcosmarxm** [comment]
Published

**Pull request closed without merging**